File: ios/Kwift/com/lightningkite/kwift/views/ios/SelectMultipleDatesMonthView.swift
//
//  SelectDayMonthView.swift
//  KwiftTemplate
//
//  Created by Joseph Ivie on 9/11/19.
//  Copyright © 2019 Joseph Ivie. All rights reserved.
//

import Foundation
import UIKit
import RxSwift


public class SelectMultipleDatesMonthView : QuickMonthView {
    var dates: MutableObservableProperty<Set<Date>> = StandardObservableProperty([]) {
        didSet {
            startup2()
            setNeedsDisplay()
        }
    }
    public override init(frame: CGRect) {
        super.init(frame: frame)
        startup2()
    }

    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        startup2()
    }

    var closers: Array<Disposable> = []
    public func startup2() {
        for close in closers {
            close.dispose()
        }
        closers.append(dates.onChange.addWeak(self) { this, value in
            this.setNeedsDisplay()
        })
    }

    override public func drawDayCell(_ ctx: CGContext, _ rect: CGRect, _ day: Date) {
        if dates.value.any({ it in day.sameDay(it) }) {
            let leftCal = Calendar.current.date(byAdding: .day, value: -1, to: day)!
            let left = dates.value.any { it in leftCal.sameDay(it) }
            let rightCal = Calendar.current.date(byAdding: .day, value: 1, to: day)!
            let right = dates.value.any { it in rightCal.sameDay(it) }
            if !left && !right {
                drawDayCellBackground(ctx: ctx, rectangle: rect, day: day, colorSet: self.selectedColorSet)
                drawDayCellText(ctx: ctx, rectangle: rect, day: day, colorSet: self.selectedColorSet)
            } else if !left && right {
                drawDayCellBackgroundStart(ctx: ctx, rectangle: rect, day: day, colorSet: self.selectedColorSet)
                drawDayCellText(ctx: ctx, rectangle: rect, day: day, colorSet: self.selectedColorSet)
            } else if left && !right {
                drawDayCellBackgroundEnd(ctx: ctx, rectangle: rect, day: day, colorSet: self.selectedColorSet)
                drawDayCellText(ctx: ctx, rectangle: rect, day: day, colorSet: self.selectedColorSet)
            } else if left && right {
                drawDayCellBackgroundMiddle(ctx: ctx, rectangle: rect, day: day, colorSet: self.selectedColorSet)
                drawDayCellText(ctx: ctx, rectangle: rect, day: day, colorSet: self.selectedColorSet)
            } else {
                drawDayCellBackground(ctx: ctx, rectangle: rect, day: day, colorSet: self.selectedColorSet)
                drawDayCellText(ctx: ctx, rectangle: rect, day: day, colorSet: self.selectedColorSet)
            }
        } else {
            drawDayCellBackground(ctx: ctx, rectangle: rect, day: day, colorSet: self.defaultColorSet)
            drawDayCellText(ctx: ctx, rectangle: rect, day: day, colorSet: self.defaultColorSet)
        }
    }

    var adding = false
    override public func onTouchDown(date: Date) -> Bool {
        adding = dates.value.none { it in date.sameDay(it) }
        return onTouchMove(date: date)
    }
    override public func onTouchMove(date: Date) -> Bool {
        if adding {
            if dates.value.none({ it in date.sameDay(it) }) {
                dates.value = dates.value.union([date])
            }
        } else {
            dates.value = dates.value.filter { it in !it.sameDay(date) }
        }
        return true
    }
    
    override public func onTouchUp(date: Date) -> Bool {
        return true
    }
}

File: android/src/main/java/com/lightningkite/khrysalis/views/android/SelectMultipleDatesMonthView.kt
package com.lightningkite.khrysalis.views.android

import android.content.Context
import android.graphics.Canvas
import android.graphics.RectF
import android.util.AttributeSet
import com.lightningkite.khrysalis.observables.Close
import com.lightningkite.khrysalis.observables.MutableObservableProperty
import com.lightningkite.khrysalis.observables.StandardObservableProperty
import com.lightningkite.khrysalis.rx.addWeak
import io.reactivex.disposables.Disposable
import java.util.*

open class SelectMultipleDatesMonthView : QuickMonthView {

    var dates: MutableObservableProperty<Set<Calendar>> = StandardObservableProperty(setOf())
        set(value) {
            field = value
            init()
            invalidate()
        }

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init()
    }

    val closers: ArrayList<Disposable> = ArrayList()
    fun init() {
        for (close in closers) {
            close.dispose()
        }
        closers += dates.onChange.addWeak(this) { self, value ->
            self.invalidate()
        }
    }

    override fun drawDayCell(canvas: Canvas, rectangle: RectF, day: Calendar) {
        if (dates.value.any { it -> day.sameDay(it) }) {
            val leftCal = (day.clone() as Calendar).apply { add(Calendar.DAY_OF_MONTH, -1) }
            val left = dates.value.any { it -> leftCal.sameDay(it) }
            val rightCal = (day.clone() as Calendar).apply { add(Calendar.DAY_OF_MONTH, 1) }
            val right = dates.value.any { it -> rightCal.sameDay(it) }

            when {
                !left && !right -> {
                    drawDayCellBackground(canvas, rectangle, selectedColorSet)
                    drawDayCellText(day, canvas, rectangle, selectedColorSet)
                }
                !left && right -> {
                    drawDayCellBackgroundStart(canvas, rectangle, selectedColorSet)
                    drawDayCellText(day, canvas, rectangle, selectedColorSet)
                }
                left && !right -> {
                    drawDayCellBackgroundEnd(canvas, rectangle, selectedColorSet)
                    drawDayCellText(day, canvas, rectangle, selectedColorSet)
                }
                left && right -> {
                    drawDayCellBackgroundMiddle(canvas, rectangle, selectedColorSet)
                    drawDayCellText(day, canvas, rectangle, selectedColorSet)
                }
                else -> {
                    drawDayCellBackground(canvas, rectangle, defaultColorSet)
                    drawDayCellText(day, canvas, rectangle, defaultColorSet)
                }
            }
            drawDayCellBackground(canvas, rectangle, selectedColorSet)
            drawDayCellText(day, canvas, rectangle, selectedColorSet)
        } else {
            drawDayCellBackground(canvas, rectangle, defaultColorSet)
            drawDayCellText(day, canvas, rectangle, defaultColorSet)
        }
    }

    var adding = false
    override fun onTouchDown(calendar: Calendar): Boolean {
        adding = dates.value.none { it -> calendar.sameDay(it) }
        onTouchMove(calendar)
        return true
    }

    override fun onTouchMove(calendar: Calendar): Boolean {
        if (adding) {
            if (dates.value.none { it -> calendar.sameDay(it) }) {
                dates.value = dates.value.plus(calendar)
            }
        } else {
            dates.value = dates.value.filter { !(it sameDay calendar) }.toSet()
        }
        return true
    }

    override fun onTouchUp(calendar: Calendar): Boolean {
        return true
    }

    private infix fun Calendar?.sameDay(other: Calendar?): Boolean =
        this != null && other != null &&
                this.get(Calendar.YEAR) == other.get(Calendar.YEAR) &&
                this.get(Calendar.MONTH) == other.get(Calendar.MONTH) &&
                this.get(Calendar.DAY_OF_MONTH) == other.get(Calendar.DAY_OF_MONTH)
}
